# misasm
Detect misassemblies based on paired-end reads

-------------------
misasm is an accurate misassembly detector designed to detect the assembly errors based on paired-end reads.
misasm can identify both misjoins and indels (i.e. insertions and deletions, e.g. >5 bp) according to the inconsistent features of aligning paired-end reads to contigs/scaffolds, such as the higher or lower coverage, short insertions, short deletions, clippings, abnormal pairing direction and larger or smaller insert size.

## Prerequisites
misasm depends on the following libraries and tools:
- [HTSlib(V1.9 or later)](https://github.com/samtools/htslib)
- g++ (v4.7 or later which supports c++11).


## Compiling misasm

The binary file can be generated by typing:
```sh
$ git clone https://github.com/zhuxiao/misasm.git
$ cd misasm/src/
$ make
```
And the binary file `misasm` will be output into the folder `misasm/src` in this package directory.


## Quick Start

Simply，misasm can be run by the command:
```sh
$ misasm scaffold.fa ecoli_50x_idba_sorted.bam
```

Then,the number of misjoins and indels will be shown:
```sh
############## Brief statistics for misassembly detection ##############
There are 17 misassemblies detected in total:
	15 misjoins
	2 indels
```

The help informartion can be shown:
```sh
Program：misasm (misassembly detector)
Version：0.1.0 (using htslib 1.9)

Usage：misasm [options] <ASM_FILE> <BAM_FILE>

Description：
	   ASM_FILE      Assembled squence file（required）
	   BAM_FILE      Coordinate-sorted BAM file（required）

Options:
     -b INT       block size [1000000]
     -s INT       detect slide size [1000]
     -u INT       detect subblock size [100]
     -g INT       gap extend region size [500]
     -a FLOAT     minimum proportion of abnormal pairing reads in misassemblies [0.3]
     -l FLOAT     minimum proportion of high-clipping reads in misassemblies [0.8]
     -i FLOAT     minimum proportion of indel reads in misassemblies [0.3]
     -n INT       minimal clipping reads supporting a misassembly [7]
     -c INT       maximal clipping region size to detect [10000]
     -o DIR       output directory [output]
     -p STR       prefix of output result files [null]
     -h           show this help message and exit
```

## Usage

```sh
$ misasm [options] <ASM_FILE> <BAM_FILE>
```

The assembled squence and sorted BAM file will be the input of misasm.

## Contact

If you have problems or some suggestions, please contact: [email](zhuxiao_hit@yeah.net) without hesitation. 

---- Enjoy !!! -----

